File: sym-aws-python-approvals/handler.py
import json
from google.protobuf import json_format
from sym.enums.service_pb2 import Service
from sym.messages.approval_pb2 import Approval
from sym.messages.expiration_pb2 import Expiration

# Approval functions get an Approval event
# Return the JSON form of an ApprovalResponse object
def handle_approve(event, context):
    try:
        approval = json_format.ParseDict(event, Approval())
        user_id = get_identity(approval.request.target.user, Service.SLACK)
        print("Target user slack id:", user_id)
        print("Target resource:", approval.request.target.resource)
        print("Target reason:", approval.request.meta.reason)
    except Exception as e:
        print("Unexpected exception:", str(e))
        return {'ok': False, 'error': str(e)}
    return {'ok': True}

# Expiration functions get an Expiration event
# Return the JSON form of an ExpirationResponse object
def handle_expire(event, context):
    try:
        expiration = json_format.ParseDict(event, Expiration())
        user_id = get_identity(expiration.target.user, Service.SLACK)
        print("Target user slack id:", user_id)
        print("Target resource:", expiration.target.resource)
    except Exception as e:
        print("Unexpected exception:", str(e))
        return {'ok': False, 'error': str(e)}
    return {'ok': True}

# Utility function to get a user ID for a specific service
# This will be in a shared Sym library eventually
def get_identity(user, service):
    for identity in user.identities:
        if identity.service == service:
            return identity.id

# Helpful for local testing
def main():
    approval = {
        'request': {
            'meta': {
                'reason': 'Access to Bucket 123'
            },
            'target': {
                'user': {
                    'identities': [{
                        'service': 'SLACK',
                        'id': 'foo@example.com'
                    }]
                },
                'resource': {
                    'id': 'my-custom-resource',
                    'service': 'CUSTOM'
                }
            }
        },
        'meta': {
            'approver': {
                'identities': [{
                    'service': 'SLACK',
                    'id': 'foo@example.com'
                }]
            }
        }
    }
    response = handle_approve(approval, '')
    print("Got Response: ", response)

if __name__ == "__main__":
    main()

File: sym-typescript-template/handler.ts
import { Handler, Context } from "aws-lambda";
import type { sym as t } from "@symops/types";
import { sym as protos } from "@symops/types";

/*
 * Create an Approval object from the incoming event and then escalate the
 * target user
 *
 * Return a response object with an error message if the action fails.
 */
export const handleApprove: Handler = async (
  event: any,
  _context: Context
): Promise<t.messages.IApprovalResponse> => {
  let approval;
  try {
    approval = protos.messages.Approval.create(event);
  } catch (error) {
    console.log(`Unable to deserialize approval: ${error}`);
    return { ok: false, error: error };
  }

  // Get the target user's slack identity (their email address)
  const identity = getIdentity(
    approval.request?.target?.user,
    protos.enums.Service.SLACK
  );
  console.log(`Target user slack id: ${identity}`);
  console.log(`Target resource service: ${approval.request?.target?.resource?.service}`);
  console.log(`Target resource id: ${approval.request?.target?.resource?.id}`);
  console.log(`Reason: ${approval.request?.meta?.reason}`);

  return { ok: true };
};

/*
 * Create an Approval object from the incoming event and then expire the
 * target user
 *
 * Return a response object with an error message if the action fails.
 */
export const handleExpire: Handler = async (
  event: any,
  _context: Context
): Promise<t.messages.IExpirationResponse> => {
  let expiration;
  try {
    expiration = protos.messages.Expiration.create(event);
  } catch (error) {
    console.log(`Unable to deserialize expiration: ${error}`);
    return { ok: false, error: error };
  }

  // Get the user's slack identity (their email address)
  const identity = getIdentity(
    expiration.target?.user,
    protos.enums.Service.SLACK
  );
  console.log(`Target user slack id: ${identity}`);
  console.log(`Target resource service: ${expiration.target?.resource?.service}`);
  console.log(`Target resource id: ${expiration.target?.resource?.id}`);

  return { ok: true };
};

/*
 * Helper function to get the user's id in a given target service
 * This will be in a shared Sym library eventually
 */
const getIdentity = (
  user: protos.models.User,
  service: protos.enums.Service
): string | null => {
  for (const identity of user.identities) {
    if (identity.service === service) {
      return identity.id;
    }
  }
};
